Trusted Types incorrectly handles invalid javascript URLs and exceptions

lukewarlow opened the pull request:
#### 1faeb75e6c430b7bd30f08fe0ef38f459f77c397
<pre>
Trusted Types incorrectly handles invalid javascript URLs and exceptions
<a href="https://bugs.webkit.org/show_bug.cgi?id=298792">https://bugs.webkit.org/show_bug.cgi?id=298792</a>

Reviewed by Darin Adler.

When the default policy returns a value that produces an invalid javsscript: URL
we now correctly send a violation report.

Likewise, if the default policy throws an exception, this also results in a violation
report being sent.

We now also correctly clear the exceptions else this can result in crashes.

Test: imported/w3c/web-platform-tests/trusted-types/navigate-to-javascript-url-008.html
* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/navigate-to-javascript-url-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/navigate-to-javascript-url-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/navigate-to-javascript-url-006.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/navigate-to-javascript-url-008-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/navigate-to-javascript-url-008.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/navigate-to-javascript-url-008.html.headers: Added.
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/navigate-to-javascript-url-csp-headers-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/support/navigation-support.js:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-navigation_06-10-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-navigation_11-15-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-navigation_21-25-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-navigation_31-35-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-navigation_36-last-expected.txt:
* Source/WebCore/bindings/js/ScriptController.cpp:
(WebCore::ScriptController::executeJavaScriptURL):
* Source/WebCore/dom/TrustedType.cpp:
(WebCore::requireTrustedTypesForPreNavigationCheckPasses):

Canonical link: <a href="https://commits.webkit.org/302543@main">https://commits.webkit.org/302543@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/362946c8738692ca54765debe53031c988e48456

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128680 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/942 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39512 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136065 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80077 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130552 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/892 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/819 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97992 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/66458 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131628 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/656 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115279 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78573 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1169 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33389 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79345 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109038 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33868 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138520 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/767 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/726 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107085 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/817 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111618 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106314 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27229 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/645 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30148 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/53147 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/830 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64085 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/693 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/748 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/777 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lukewarlow commented on LayoutTests/imported/w3c/web-platform-tests/trusted-types/navigate-to-javascript-url-008.html:
```diff
@@ -0,0 +1,20 @@
+<!DOCTYPE html>
```
Exported at https://github.com/web-platform-tests/wpt/pull/55736

darinadler commented on LayoutTests/imported/w3c/web-platform-tests/trusted-types/support/navigation-support.js:
```diff
@@ -59,7 +59,9 @@ async function setLocationToJavaScriptURL(defaultpolicy) {
-          await new Promise(resolve => requestIdleCallback(resolve));
+            await new Promise(resolve => requestIdleCallback(resolve));
+        } else {
+          await new Promise(resolve => requestAnimationFrame(_ => requestAnimationFrame(resolve)));
```
Inconsistent indentation here
```suggestion
          await new Promise(resolve => requestIdleCallback(resolve));
        } else {
          await new Promise(resolve => requestAnimationFrame(_ => requestAnimationFrame(resolve)));
```

darinadler commented on Source/WebCore/dom/TrustedType.cpp:
```diff
@@ -289,25 +289,27 @@ ExceptionOr<String> requireTrustedTypesForPreNavigationCheckPasses(ScriptExecuti
-        return WTFMove(std::get<Exception>(convertedScriptSource));
+    if (std::holds_alternative<Exception>(convertedScriptSource)) {
+        throwScope.clearException();
+    } else if (!std::holds_alternative<std::monostate>(convertedScriptSource)) {
```
WebKit coding style omits braces for one-line if statement bodies:
```suggestion
    if (std::holds_alternative<Exception>(convertedScriptSource))
        throwScope.clearException();
    else if (!std::holds_alternative<std::monostate>(convertedScriptSource)) {
```

darinadler reviewed (approved):
This looks right to me, although you _might_ want to wait for comments from the other reviewers.

darinadler commented on Source/WebCore/dom/TrustedType.cpp:
```diff
@@ -289,25 +289,27 @@ ExceptionOr<String> requireTrustedTypesForPreNavigationCheckPasses(ScriptExecuti
-            return Exception { ExceptionCode::TypeError, makeString("This assignment requires a "_s, trustedTypeToString(expectedType)) };
-
-        return String(urlString);
+        if (newURL.isValid())
```
This is not new but it's worth thinking through: I'm not sure how this validity check could ever be false. We certainly don't have a test covering that case!

I also suspect the conversion to URL and back to a string does not accomplish much; it would be nice to know why we do it, and test that behavior, in case understanding that inspires us to write the code differently. The main purpose of the URL class is parsing and there is no real parsing of JavaScript URLs that I am aware of.

lukewarlow replied:
It can happen if the default policy returns a value that's not able to be parsed as a valid URL when put into a javascript URL.

e.g. "//make:invalid/"

This is prepended with "javascript:" and that makes newURL invalid. You can test this with `data:text/html,<script>alert(URL.canParse("javascript://make:invalid/"))</script>`

We have a test for this in the form of https://wpt.fyi/results/trusted-types/navigate-to-javascript-url-004.html

darinadler replied:
OK. I'd also like to understand if the round tripping through URL is needed or just the validity checking. But this is not urgent, just feels a little loose to me.

lukewarlow committed:
Trusted Types incorrectly handles invalid javascript URLs and exceptions
https://bugs.webkit.org/show_bug.cgi?id=298792

Reviewed by Darin Adler.

When the default policy returns a value that produces an invalid javsscript: URL
we now correctly send a violation report.

Likewise, if the default policy throws an exception, this also results in a violation
report being sent.

We now also correctly clear the exceptions else this can result in crashes.

Test: imported/w3c/web-platform-tests/trusted-types/navigate-to-javascript-url-008.html
* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/navigate-to-javascript-url-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/navigate-to-javascript-url-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/navigate-to-javascript-url-006.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/navigate-to-javascript-url-008-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/navigate-to-javascript-url-008.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/navigate-to-javascript-url-008.html.headers: Added.
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/navigate-to-javascript-url-csp-headers-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/support/navigation-support.js:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-navigation_06-10-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-navigation_11-15-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-navigation_21-25-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-navigation_31-35-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-navigation_36-last-expected.txt:
* Source/WebCore/bindings/js/ScriptController.cpp:
(WebCore::ScriptController::executeJavaScriptURL):
* Source/WebCore/dom/TrustedType.cpp:
(WebCore::requireTrustedTypesForPreNavigationCheckPasses):

Canonical link: https://commits.webkit.org/302543@main
